--- urdf_model/include/urdf_model/joint.h
+++ urdf_model/include/urdf_model/joint.h
@@ -162,6 +162,22 @@
   };
 };
 
+class Joint4LinksDescription
+{
+public:
+    Joint4LinksDescription() { this->clear(); };
+    double r1, r2;
+    double d;
+    double l;
+    double theta;
+
+    void clear()
+    {
+      r1 = 0.0;    r2 = 0.0;
+      d  = 0.0;    l  = 0.0;
+      theta = 0.0;
+    }
+};
 
 class Joint
 {
@@ -172,7 +188,7 @@
   std::string name;
   enum
   {
-    UNKNOWN, REVOLUTE, CONTINUOUS, PRISMATIC, FLOATING, PLANAR, FIXED
+    UNKNOWN, REVOLUTE, CONTINUOUS, PRISMATIC, FLOATING, PLANAR, FIXED, FOURLINKS
   } type;
 
   /// \brief     type_       meaning of axis_
@@ -210,6 +226,8 @@
   /// Option to Mimic another Joint
   boost::shared_ptr<JointMimic> mimic;
 
+  boost::shared_ptr<Joint4LinksDescription> description4l;
+
   void clear()
   {
     this->axis.clear();
